### DISCUSSION TITLE
GrSim上でマイコンプログラムを動かすためのインタフェースについて

### DISCUSSION THREAD

**Discussion by HansRobo** [ibis-ssl/crane#370] [Ideas]
GrSimと実機の動きがかなり違ったりしていて、結構バグの原因になっている。
マザー基板のプログラムのインタフェースを抽象化して、GrSim内で動かせるようにすることで効率的な開発体制を目指したい。

https://github.com/ibis-ssl/G474_Orion_main
https://github.com/ibis-ssl/grSim

**Comment by HansRobo** [comment]
## 参考ソースコード

- https://github.com/ibis-ssl/grSim/blob/master/include/robot.h
- https://github.com/ibis-ssl/grSim/blob/master/src/robot.cpp

## Input(シミュレータ外から)

### 現状
https://github.com/ibis-ssl/grSim/blob/master/src/proto/grSim_Commands.proto
- ロボット座標系vx
- ロボット座標系vy
- 角速度
- キックパワー
- キックモード
- ドリブルパワー
- （各車輪の速度制御モードもある）

### 理想

CM4を経由して入力されるAIからのパケットをそのまま受け取って処理させたい。  
AI側からは送信先を変えるだけで実機向けと全く同じパケットを送りたい 

## Input(シミュレータ内から)

位置などはそのままでも取得できるはず。  
角速度などは計算する必要があるかも？  
ジャイロやマウスセンサなどの各種センサをシミュレートする実装が必要かも？

## Output(シミュレータ外へ)

### 現状

https://github.com/ibis-ssl/grSim/blob/master/src/proto/grSim_Robotstatus.proto
- ボールセンサON/OFF
- キッカーのモード情報

### 理想

実機が出力しているフィードバック情報全て

## Output(シミュレータ内)

### 現状

受け取った速度・角速度から車輪回転速度を計算してそのまま回してそう
キッカーは受け取った速度で動作。ドリブラはパワー関係なく剛体接着してそう。

### 理想

AIパケットで得た目標位置や自己位置を元に自分で計算して車輪を回してほしい。
キッカーやドリブラのリアルさももうちょっと上げたい。

**Comment by HansRobo** [comment]
# マイコンプログラムをGrSimで動かすために大事そうなこと

- メインプログラムで直にHALの関数を叩かない( OR GrSIm側で空のHAL関数を作っておく)
- C++使いたい...
- 割り込みをGrSIｍ側でどうにか表現したい（簡単に使えるタイマーがQtにあればいいが....）
- どの単位で埋め込むか考える
  - main.cを適当に配置してビルドしたら動くようにするとか？

**Reply by ChikuwaRobo** [reply]
- 今はロボット側の機体制御は500Hzで回っていて、今後1kHzに上がるかもしれない
- シミュレータ側もその速度で回すのはしんどい？
- 機体の加速度･速度制御だけ別ファイルにわけてて、そこだけGrSim側に取り込むので良さそう
- ロボット側でやっている加速度制御(加速度の異方性)が再現できるようになるだけでも、あるていど目標は達成できそう

**Reply by HansRobo** [reply]
> シミュレータ側もその速度で回すのはしんどい？

正直やってみないとわからん、という感想だがシミュレーションは現実時間で動かさなくても良いので、計算がきつければ速度を落とせば良いかも。

> 機体の加速度･速度制御だけ別ファイルにわけてて、そこだけGrSim側に取り込むので良さそう

ミニマルスタートでこれくらいで始めても良いかもしれん。
ただ、自己位置推定周りを統合してVisionとの兼ね合いなどがシミュレーションできるようになるのはかなり捨てがたいのでゆくゆくは出来るようにしたいところ。

> ロボット側でやっている加速度制御(加速度の異方性)が再現できるようになるだけでも、あるていど目標は達成できそう

ロボットの車輪の角度の変更や重心の設定などである程度再現できそうな気がする。

**Comment by ChikuwaRobo** [comment]
# ロボット側で明示的に使っている情報
- AI指令値
- 現在速度(odom)
- 現在座標(odom)
- ジャイロによるヨー角度


# シミュレーター上でロボットの挙動を再現するにあたって
ジャイロによる角度はシミュレータ上の理論値を使ってしまっても支障はなさそう。

odomはかなり厄介で、理論値を使ってしまうと実機特有の挙動が再現できなくなってしまう。
ところが、本気でやろうとした場合
- モーター特性
- モーターやドライバ都合の出力リミット
- 加速度による各輪荷重
- フィールドとの摩擦

のシミュレーションをした上で、odomに使用しているタイヤの回転数を得なければならない。
ここまでやるのは流石にオーバーな気がするので、目的に合うだけの簡略化された実装にしたい。
(とはいえどうしたもんかなぁ)

**Reply by HansRobo** [reply]
@KANAIHIROYUKI

詳しくは実装を調べないといけないが、
シミュレーションは基本的にODEを使って行われていて、車輪やフィールドはその中でシミュレーションアイテム化されているはずなので、
以下の2点は既に実装されているはず。

- フィールドとの摩擦
- 加速度による各輪荷重

**Reply by ChikuwaRobo** [reply]
なるほど、じゃあわりとなんとかなる気がしてきた(そのへんをいい感じに調整する必要はあるにせよ)
あと出力リミットもある(追記しといた)ので、ロボット側からの出力だけ再現してあげればそれなりにリアルに近づけられそうかなぁ

**Comment by HansRobo** [comment]
# 第1弾実装

議論した感じ、やることが色々あるのでまずはやることを絞って実装したほうが良いと思った。

- GrSim用ではない、ibisのAIパケットを直で受け取る
- 加速度・速度制御の埋め込み
- ロボット車輪角度と重心の設定

くらいがちょうどいいかも

@KANAIHIROYUKI どうでしょう？

**Reply by ChikuwaRobo** [reply]
とりあえずそれで良いと思う

odomの理論値と現実のズレを再現する手段については、GrSim(ODE)のシミュレーションを使うよりは、
別でノイズと言うかエラーを乗せる仕組みにしたほうがいいかもしれないので、ロボット車輪角度とかの話は一旦置いておいて良いかも。
まずは動かしてみたい

**Reply by HansRobo** [reply]
> まずは動かしてみたい

わかる。
やるぞやるぞやるぞやるぞやるぞやるぞやるぞやるぞやるぞやるぞやるぞやるぞ〜〜〜〜〜〜！！！！！！！

**Reply by HansRobo** [reply]
@KANAIHIROYUKI 以下の情報がほしい

> 加速度・速度制御の埋め込み

これのソースコード情報

> ロボット車輪角度と重心の設定

これの情報

**Reply by ChikuwaRobo** [reply]
実作業はそっちでやる？


> ソースコード情報

現状、マイコン側がだいぶとっ散らかったまま(大会直前にいじったまんまだし)だから、リファクタして別ファイルに分けたのを参照してもらう感じで良いかな？


> ロボット車輪角度と重心の設定

ここはマイコン側でやってるodom計算をGrSim側でどこまでやるかという話になる(車輪角度がスリップなしの理論値で出てくるなら、シミュレータ側の座標をそのまま突っ込んでも同じになる)ので、今はいいかな？
重心は今はまだ考慮しなくて良い(なぜならロボット側が重心による荷重の偏りを考慮した制御を実装できていないので)

**Reply by HansRobo** [reply]
実作業はどちらでも大丈夫だけど、自分が今までGrSimを少し改造してるので自分がやると良いかなと思った。
ちくわはマイコン側のリファクタに専念してもらうのが良いかも。

> ソースコード情報

リファクタ、頼んだ。いい感じになったらまた教えてほしい。

> ロボット車輪角度と重心の設定

これちょっと語弊があったんだけど、「車輪角度」というのは「ロボットを真上から見た車輪の向いている角度」のことを指していた。
「車輪、つまりモーターが出力する角度」の方は現状でもODEがある程度スリップさせてくれる気がする。
重心考慮なし、了解しました。

**Reply by ChikuwaRobo** [reply]
理解。
モーター角度は正面を0にして、
+60,-60,+135,-135

**Reply by HansRobo** [reply]
あざす

**Reply by HansRobo** [reply]
リファクタ済みのマイコンソースコード
https://github.com/ibis-ssl/G474_Orion_main/blob/main/Core/Src/robot_control.c

**Reply by HansRobo** [reply]
作業中
https://github.com/ibis-ssl/grSim/pull/3

**Reply by ChikuwaRobo** [reply]
@HansRobo 
robot_control.c、最近機体制御側をいろいろ見直してる最中だからめちゃ内容変わっちゃってるんで、どっかで更新よろ

**Reply by HansRobo** [reply]
承知。先日もスペルミス直したところだしちょうどどこかで更新しようと思ってたところ

**Reply by ChikuwaRobo** [reply]
と言いつつまだいろいろ変更入れそうだわ･･･

**Reply by HansRobo** [reply]
言うてコピペするだけだからそんな手間でもない
ファイルの中からインクルード含めてマイコン関連を排除できればファイルのコピーだけで使い回せるようになるからもっと楽になるな〜（チラッ）

**Reply by ChikuwaRobo** [reply]
わかるっピねぇ～～～(とはいえミッチミチに密結合になってるのでだいぶ大変そう)

**Reply by HansRobo** [reply]
マイコン側の制御はまだ使えていないけど、実機と同じパケットをgrSim側で受け取ってロボットを動かすのはできた

https://github.com/ibis-ssl/crane/assets/12395284/a65ad60a-088d-4d52-98f3-ca71883fbcad

**Reply by ChikuwaRobo** [reply]
神～～

**Comment by HansRobo** [comment]
本プロジェクトはクローズしました。
理由は、本当に実機に似せようとすると実装が大変なことや、意味わからんところでパラチュンが発生するからです。
